### PR TITLE
Disallow `^` in `def` command names

### LINKED ISSUE #7273: Possible to create `def`s that start with `^` (thus preventing external command access)
### Describe the bug

See title.

### How to reproduce

```
〉def ^cat [] { 'Bah!' }
〉^cat
Bah!
```

### Expected behavior

I expect the ^ to be a protected syntactic element that can't be overridden by someone making a nefarious `def`.

### Screenshots

_No response_

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.71.0                                          |
| branch             |                                                 |
| commit_hash        | 6cc4ef6c706ebbdc5b995013e0b00f77f7a4dba7        |
| build_os           | windows-x86_64                                  |
| build_target       | x86_64-pc-windows-msvc                          |
| rust_version       | rustc 1.65.0 (897e37553 2022-11-02)             |
| rust_channel       | stable-x86_64-pc-windows-msvc                   |
| cargo_version      | cargo 1.65.0 (4bc8f24d3 2022-10-20)             |
| pkg_version        | 0.71.0                                          |
| build_time         | 2022-11-08 18:03:36 +00:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  |                                                 |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by webbedspace**
# Description

Closes #7273.

Also slightly edits/tidies up parser.rs.

# User-Facing Changes

`^` is now forbidden in `def` and `def-env` command names. EDIT: also `alias`.

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

### CODE REVIEW

File: crates/nu-parser/src/parser.rs [hunk 1]
```diff
@@ -3291,8 +3294,22 @@ pub fn parse_signature_helper(
...
+                                ($name:expr) => {
+                                    if !is_variable($name) {
+                                        error = error.or_else(|| {
+                                            Some(ParseError::Expected(
+                                                "valid variable name".into(),
+                                                span,
+                                            ))
+                                        })
+                                    }
+                                };
```
Comment (sholderbach) [inline crates/nu-parser/src/parser.rs:3307]: Not convinced that this macro is necessary for this few usages. I would rather have a function that emits a useful error and has documentation to encourage usage in other places if the need arises.

**Review by sholderbach — COMMENTED** [review]
Thanks thats a very good point! Also great to have a few more tests and some additional comments in the parser spaghetti.

File: crates/nu-parser/src/parser.rs (hunk 1 above)
Comment (webbedspace) [reply crates/nu-parser/src/parser.rs]: I don't know how you'd rewrite this to be a function given that the point of it is to dedup the `error = error.or_else()` code structure, not just produce the `ParseError::Expected`. I also think 6 uses is about enough for an 8-line block. But w/e.

**Review by rgwood — APPROVED** [review]
This is a nice improvement, thanks.

**Pull request merged**